### ISSUE TITLE
Running eksctl in Docker results with ExpiredToken error

### ISSUE THREAD

**Issue by darnone** [weaveworks/eksctl#3514]
We are trying to deploy an eks cluster.  Running from shell scripts natively works with no problems. 

For automation we are using Docker to hold all of the deployment tools - eksctl, kubectl, etc. The deployment is executed with a `docker-compose up` on an ec2 bastion server in a different account that is routed to a new account with no vpc. The `network_mode` in the `docker-compose.yaml` file is host. The image being build for deployment uses a base docker image with all of the tools and adds to that the deployment code and mounts it iat /opt inside the container. The docker-compose file simply calls a script inside the container which in turn calls the automation code. 

When the control plane starts to create  - I can see the normal on cloudformation messages waiting on cloudformation, but after a short time I get a token  ExpiredToken errors:
```
 |[0m 2021-04-01 18:04:18 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:04:36.9430976Z [36mbuilder_1  |[0m 2021-04-01 18:04:36 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:04:36.9611752Z [36mbuilder_1  |[0m 2021-04-01 18:04:36 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:36.9612980Z [36mbuilder_1  |[0m 	status code: 403, request id: 38342986-a67c-4f39-b088-43f31f74b0e7) from cloudformation/DescribeStacks - will retry after delay of 55.520858ms
2021-04-01T18:04:37.0254704Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.0255987Z [36mbuilder_1  |[0m 	status code: 403, request id: d2f39c17-3782-4aeb-92a4-04149d525eb4) from cloudformation/DescribeStacks - will retry after delay of 92.551074ms
2021-04-01T18:04:37.1222739Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.1223981Z [36mbuilder_1  |[0m 	status code: 403, request id: 65bbdae8-66f3-4b2d-8216-4e31238704de) from cloudformation/DescribeStacks - will retry after delay of 219.677144ms
2021-04-01T18:04:37.3476133Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
```

sometimes the error occurs during the nodegroup creation:

```
2021-04-01T18:02:29.2392992Z [36mbuilder_1  |[0m 2021-04-01 18:02:29 [ℹ]  deploying stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:02:29.2394240Z [36mbuilder_1  |[0m 2021-04-01 18:02:29 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:02:45.6752524Z [36mbuilder_1  |[0m 2021-04-01 18:02:45 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:03:02.4878842Z [36mbuilder_1  |[0m 2021-04-01 18:03:02 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:03:22.1089353Z [36mbuilder_1  |[0m 2021-04-01 18:03:22 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:03:38.7797990Z [36mbuilder_1  |[0m 2021-04-01 18:03:38 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:03:58.7803073Z [36mbuilder_1  |[0m 2021-04-01 18:03:58 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:04:18.1479986Z [36mbuilder_1  |[0m 2021-04-01 18:04:18 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:04:36.9430976Z [36mbuilder_1  |[0m 2021-04-01 18:04:36 [ℹ]  waiting for CloudFormation stack "eksctl-eks-nrg-new1-cluster-test-account1-nodegroup-eks-nrg-new1-nodegroup"
2021-04-01T18:04:36.9611752Z [36mbuilder_1  |[0m 2021-04-01 18:04:36 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:36.9612980Z [36mbuilder_1  |[0m 	status code: 403, request id: 38342986-a67c-4f39-b088-43f31f74b0e7) from cloudformation/DescribeStacks - will retry after delay of 55.520858ms
2021-04-01T18:04:37.0254704Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.0255987Z [36mbuilder_1  |[0m 	status code: 403, request id: d2f39c17-3782-4aeb-92a4-04149d525eb4) from cloudformation/DescribeStacks - will retry after delay of 92.551074ms
2021-04-01T18:04:37.1222739Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.1223981Z [36mbuilder_1  |[0m 	status code: 403, request id: 65bbdae8-66f3-4b2d-8216-4e31238704de) from cloudformation/DescribeStacks - will retry after delay of 219.677144ms
2021-04-01T18:04:37.3476133Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.3477334Z [36mbuilder_1  |[0m 	status code: 403, request id: 77b17fcd-ea76-47ce-8981-dd09e6a6b0e8) from cloudformation/DescribeStacks - will retry after delay of 410.791472ms
2021-04-01T18:04:37.7631293Z [36mbuilder_1  |[0m 2021-04-01 18:04:37 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:37.7633629Z [36mbuilder_1  |[0m 	status code: 403, request id: fbf520f8-62b6-4b8a-a5e7-adb36b155372) from cloudformation/DescribeStacks - will retry after delay of 710.164976ms
2021-04-01T18:04:38.4781977Z [36mbuilder_1  |[0m 2021-04-01 18:04:38 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:38.4783196Z [36mbuilder_1  |[0m 	status code: 403, request id: 599a05dd-81eb-44ef-a386-05e818719f41) from cloudformation/DescribeStacks - will retry after delay of 1.000753504s
2021-04-01T18:04:39.4838587Z [36mbuilder_1  |[0m 2021-04-01 18:04:39 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:39.4839762Z [36mbuilder_1  |[0m 	status code: 403, request id: f2add994-050b-4a97-94c9-170ca37e9586) from cloudformation/DescribeStacks - will retry after delay of 3.565630976s
2021-04-01T18:04:43.0578464Z [36mbuilder_1  |[0m 2021-04-01 18:04:43 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:43.0579911Z [36mbuilder_1  |[0m 	status code: 403, request id: 84d40207-53ec-403e-b78e-e8b0303bcdb0) from cloudformation/DescribeStacks - will retry after delay of 6.359421824s
2021-04-01T18:04:49.4351126Z [36mbuilder_1  |[0m 2021-04-01 18:04:49 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:49.4352322Z [36mbuilder_1  |[0m 	status code: 403, request id: 406f7dc6-84eb-4f1a-a76c-4ea154405130) from cloudformation/DescribeStacks - will retry after delay of 9.693614592s
2021-04-01T18:04:59.1509581Z [36mbuilder_1  |[0m 2021-04-01 18:04:59 [!]  retryable error (ExpiredToken: The security token included in the request is expired
2021-04-01T18:04:59.1511268Z [36mbuilder_1  |[0m 	status code: 403, request id: 847087fc-271d-4bcd-b1eb-b7dcc3f946e6) from cloudformation/DescribeStacks - will retry after delay of 20.796567552s
```

from another thread I found something about [increasing the hop limit](https://github.com/weaveworks/eksctl/issues/1979) so executed the following against the ec2 instance:

```
To be able to run eksctl in a container and have it successfully authenticate, you will need to have Docker use the host network (to reduce the request's hop count) or increase the max allowed hops to 2 like so:

aws ec2 modify-instance-metadata-options \
    --instance-id i-1234567898abcdef0 \
    --http-put-response-hop-limit 2 \
    --http-endpoint enabled
```
But I am still getting the error.

If I run my scripts from my laptop, deployment works
If I run the deployment natively on the bastion host it works
If I run the deployment in Azure Devops it time out - difference is in is executing in container using docker-compose
If I run docker-compose up on the bastion host manually it times works. So is this hit and miss?

From the thread above the alternative is to use the host networking. I have network_mode: host set the in the compose file. Isn't that setting the networking to host?  If not how should I execute this docker call so that host networking is used?

The help I am looking for is how to resolve the error.  The docker compose is attached.
[docker-compose.txt](https://github.com/weaveworks/eksctl/files/6245780/docker-compose.txt). 

I am using eksctl 0.41.0

**Comment by Callisto13** [comment]
Hi @darnone could you check the session expiration of the `~/.aws/credentials` that the docker container is using?

**Comment by darnone** [comment]
We wrote a python script that assumes the existing role for 3600 seconds an injects the credentials as env variable back into the existing process the problem is that they are lasting for 3600 seconds, however, if I inject them often enough and frequently, we can get through the deployment. It is not elegant but it works.  Not sure why the creds aren't good for an hour but that is probably a question for Amazon.

**Comment by darnone** [comment]
Claudia,  Awesome bird on your hand there. I worked with AWS support on this. The issue originates from assuming a role from a role.  When using an ADO AWS Shell, you provide it the AWS IAM user (and/or role to assume).  In the fine print of role assumption, it says the role max duration is only good for 15 minutes. There were two ways to resolve this - set a  parameter`aws.rolecredential.maxduration` in the release or pipeline definition to extend the default or assume the role in the script. We choose the latter and it worked like a charm. That being said, we have decided to shift our deployments to GitHub Actions runner which we can run as pod in the Kubernetes cluster itself.  We think that is a much better solution.
